src/IRMatch.h

```diff
@@ -1839,6 +1881,109 @@ HALIDE_ALWAYS_INLINE auto ramp(A &&a, B &&b, C &&c) noexcept -> RampOp<decltype(
     return {pattern_arg(a), pattern_arg(b), pattern_arg(c)};
 }
 
+template<typename... Args>
+struct VectorIntrinOp {
+    struct pattern_tag {};
+    const std::string &intrin_name;
+    std::tuple<Args...> args;
+
+    static constexpr uint32_t binds = bitwise_or_reduce((bindings<Args>::mask)...);
+
+    constexpr static IRNodeType min_node_type = IRNodeType::VectorIntrinsic;
+    constexpr static IRNodeType max_node_type = IRNodeType::VectorIntrinsic;
+    constexpr static bool canonical = and_reduce((Args::canonical)...);
+
+    template<int i,
+             uint32_t bound,
+             typename = typename std::enable_if<(i < sizeof...(Args))>::type>
+    HALIDE_ALWAYS_INLINE bool match_args(int, const VectorIntrinsic &v, MatcherState &state) const noexcept {
+        using T = decltype(std::get<i>(args));
+        return (std::get<i>(args).template match<bound>(*v.args[i].get(), state) &&
+                match_args<i + 1, bound | bindings<T>::mask>(0, v, state));
+    }
+
+    template<int i, uint32_t binds>
+    HALIDE_ALWAYS_INLINE bool match_args(double, const VectorIntrinsic &v, MatcherState &state) const noexcept {
+        return true;
+    }
+
+    template<uint32_t bound>
+    HALIDE_ALWAYS_INLINE bool match(const BaseExprNode &e, MatcherState &state) const noexcept {
+        if (e.node_type != IRNodeType::VectorIntrinsic) {
+            return false;
+        }
+        const VectorIntrinsic &v = (const VectorIntrinsic &)e;
+        return (v.name == intrin_name && match_args<0, bound>(0, v, state));
+    }
+
+    template<int i,
+             typename = typename std::enable_if<(i < sizeof...(Args))>::type>
+    HALIDE_ALWAYS_INLINE void print_args(int, std::ostream &s) const {
+        s << std::get<i>(args);
+        if (i + 1 < sizeof...(Args)) {
+            s << ", ";
+        }
+        print_args<i + 1>(0, s);
+    }
+
+    template<int i>
+    HALIDE_ALWAYS_INLINE void print_args(double, std::ostream &s) const {
+    }
+
+    HALIDE_ALWAYS_INLINE
+    void print_args(std::ostream &s) const {
+        print_args<0>(0, s);
+    }
+
+    HALIDE_ALWAYS_INLINE
+    Expr make(MatcherState &state, halide_type_t type_hint) const {
+        std::vector<Expr> r_args(sizeof...(Args));
+        // TODO(rootjalex): How do we do type hints for the args?
+        // TODO(rootjalex): Is there a way to do basically an unrolled
+        // loop of the below? this is ugly.
+        // Supposedly C++20 will have constexpr std::transform, perhaps
```

Comment:
It will likely be a *long* while before Halide can rely on C++20 being available.

Reply:
Gotcha. Is there a reason why? I thought we just upgraded to C++17, so I figured C++20 wouldn't be that far off.

Reply:
And thanks for taking a look at this!

Reply:
Anecdotally, C++20 is getting adopted at a *much* slower pace than C++11 or C++17 did. (It's not clear that there's even a timetable to adopt it inside Google.) We'd have to be confident that ~all significant Halide customers are ready to make that commitment.

Reply:
Gotcha, thanks for the explanation! I guess I'll stick to the ugly `if constexpr` expressions :')